File: data-utility/scheduled_tasks/backup_exp_video.m
% INCREMENTAL_VIDEO_BACKUP
% - Scans camNN_video_r folders under each subject
% - Copies only NEW or CHANGED video files into a time-stamped batch folder
% - Preserves experiment/subject structure (experiment_##/__parent_child/...)
% - Updates a CSV manifest so future runs are incremental
%
% Assumes your helper functions exist on path:
%   cIDs(exp_ids), get_subject_info(sub_id), get_subject_dir(sub_id), sub2exp(sub_id)

%% ====== USER CONFIG ======
exp_ids        = [12, 15, 351, 353, 361, 362, 363, 70:75, 91, 27, 77:79, 65:69, 58, 59, 96];
videoExts      = {'.mp4', '.avi', '.mov', '.mkv', '.m4v'}; % case-insensitive
camDirRegex    = '^cam\d{2}_video_r$';                     % camera folder naming
% Where to store the incremental backups (persistent location):
dest_root      = 'Y:\multiwork_active_exp_backup\video_backup_incremental';

% Optional: dry run (true = simulate, print what would be copied)
dry_run = false;
%% =========================

if ~exist(dest_root, 'dir'); mkdir(dest_root); end
manifest_path = fullfile(dest_root, 'manifest.csv');
batch_dir     = fullfile(dest_root, datestr(datetime('now'), 'yyyymmdd_HHMMSS'));
if ~exist(batch_dir, 'dir'); mkdir(batch_dir); end

% Load existing manifest (if any)
manifest = load_manifest(manifest_path);

% Collect candidate files from all subjects
sub_ids = cIDs(exp_ids);

fprintf('[%s] Scanning subjects...\n', datestr(now));
to_copy = table('Size',[0 5], ...
    'VariableTypes', {'string','double','double','string','string'}, ...
    'VariableNames', {'rel_path','size','mtime','src','dst'});

for s = 1:numel(sub_ids)
    sub_id  = sub_ids(s);
    try
        sub_info = get_subject_info(sub_id);
        sub_dir  = get_subject_dir(sub_id);
        if ~exist(sub_dir, 'dir')
            warning('Subject folder does not exist: %s', sub_dir);
            continue;
        end

        exp_num  = sub2exp(sub_id);
        rel_prefix = fullfile(sprintf('experiment_%d', exp_num), sprintf('__%d_%d', sub_info(3), sub_info(4)));

        % find all camNN_video_r folders directly under subject
        d = dir(sub_dir);
        candDirs = d([d.isdir]);
        for k = 1:numel(candDirs)
            nameK = candDirs(k).name;
            if ismember(nameK, {'.','..'}); continue; end
            if ~isempty(regexp(nameK, camDirRegex, 'once'))
                cam_root = fullfile(sub_dir, nameK);
                % enumerate video files recursively
                files = list_files_with_ext(cam_root, videoExts);
                for i = 1:height(files)
                    src = files.path(i);
                    % rel from cam_root
                    rel_under_cam = erase(src, cam_root);
                    if startsWith(rel_under_cam, filesep), rel_under_cam = extractAfter(rel_under_cam, 1); end
                    rel_path = fullfile(rel_prefix, nameK, rel_under_cam);

                    file_info = dir(src);
                    this_size = double(file_info.bytes);
                    this_mtime = datenum(file_info.datenum);

                    % decide if this file is new or changed
                    idx = find(manifest.rel_path == string(rel_path), 1);
                    is_new = isempty(idx);
                    is_changed = false;
                    if ~is_new
                        % Changed if size differs or file mod-time is newer
                        is_changed = (manifest.size(idx) ~= this_size) || (manifest.mtime(idx) < this_mtime);
                    end

                    if is_new || is_changed
                        dst = fullfile(batch_dir, rel_path);
                        to_copy = [to_copy; {string(rel_path), this_size, this_mtime, string(src), string(dst)}]; %#ok<AGROW>
                    end
                end
            end
        end
    catch ME
        warning('Error scanning subject %d: %s', sub_id, ME.message);
        continue;
    end
end

% Deduplicate rows (defensive) and ensure parent dirs exist
if ~isempty(to_copy)
    [~, ia] = unique(to_copy.rel_path, 'stable');
    to_copy = to_copy(ia, :);
end

fprintf('[%s] Files to copy: %d\n', datestr(now), height(to_copy));

% Copy changed/new files
n_ok = 0; n_fail = 0;
for i = 1:height(to_copy)
    dstDir = fileparts(to_copy.dst(i));
    if ~dry_run && ~exist(dstDir, 'dir'); mkdir(dstDir); end
    try
        if ~dry_run
            copyfile(to_copy.src(i), to_copy.dst(i), 'f');
        end
        n_ok = n_ok + 1;

        % Update manifest row for this file
        idx = find(manifest.rel_path == to_copy.rel_path(i), 1);
        if isempty(idx)
            manifest = [manifest; table(to_copy.rel_path(i), to_copy.size(i), to_copy.mtime(i), ...
                                        'VariableNames', {'rel_path','size','mtime'})]; %#ok<AGROW>
        else
            manifest.size(idx)  = to_copy.size(i);
            manifest.mtime(idx) = to_copy.mtime(i);
        end
    catch ME
        n_fail = n_fail + 1;
        warning('Copy failed: %s -> %s (%s)', to_copy.src(i), to_copy.dst(i), ME.message);
    end
end

% Save manifest if not dry-run
if ~dry_run
    save_manifest(manifest_path, manifest);
end

fprintf('[%s] Incremental backup done. Copied OK: %d, Failed: %d, Batch folder: %s\n', ...
    datestr(now), n_ok, n_fail, batch_dir);

% function

%% ===================== helpers =====================

function files = list_files_with_ext(root, exts)
% Return table with column "path" (string) of files under root that have one of exts.
    exts = lower(string(exts));
    files = table(string.empty(0,1), 'VariableNames', {'path'});

    stack = {root};
    while ~isempty(stack)
        cur = stack{end}; stack(end) = [];
        L = dir(cur);
        for i = 1:numel(L)
            if L(i).isdir
                if ~ismember(L(i).name, {'.','..'})
                    stack{end+1} = fullfile(cur, L(i).name); %#ok<AGROW>
                end
            else
                [~,~,e] = fileparts(L(i).name);
                if any(strcmpi(lower(string(e)), exts))
                    files.path(end+1,1) = string(fullfile(cur, L(i).name)); %#ok<AGROW>
                end
            end
        end
    end
end

function T = load_manifest(p)
% Load CSV with columns: rel_path,size,mtime
    if exist(p, 'file')
        T = readtable(p, 'TextType', 'string');
        if ~all(ismember({'rel_path','size','mtime'}, T.Properties.VariableNames))
            error('Manifest missing required columns.');
        end
        % Ensure types
        T.rel_path = string(T.rel_path);
        T.size     = double(T.size);
        T.mtime    = double(T.mtime);
    else
        T = table(string.empty(0,1), double.empty(0,1), double.empty(0,1), ...
            'VariableNames', {'rel_path','size','mtime'});
    end
end

function save_manifest(p, T)
% Save CSV manifest
    writetable(T, p);
end

File: data-utility/scheduled_tasks/backup_exp_video.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
INCREMENTAL VIDEO BACKUP
- Scans all experiment_* folders in MULTIWORK_ROOT.
- For each experiment, finds subjects inside 'included' matching '__YYYYMMDD_XXXXX'.
- Under each subject, finds camera folders matching 'camNN_video_r'.
- Recursively copies only NEW or CHANGED video files (by size/mtime) to a time-stamped batch.
- Updates persistent CSV manifest so future runs are incremental.

Tested on Windows-style paths. Requires Python 3.9+.
"""

import os
import re
import csv
import sys
import shutil
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Iterable, Dict, Tuple, List

# =============== CONFIG ===============
MULTIWORK_ROOT = Path(r"M:\ ").resolve()     # root containing experiment_* folders
MULTIWORK_ROOT = Path(str(MULTIWORK_ROOT).strip())

DEST_ROOT      = Path(r"Y:\multiwork_active_exp_backup\video_backup_incremental").resolve()
VIDEO_EXTS     = ['.mp4', '.avi', '.mov', '.mkv', '.m4v']  # case-insensitive
CAM_DIR_REGEX  = re.compile(r"^cam\d{2}_video_r$", re.IGNORECASE)
EXP_REGEX      = re.compile(r"^experiment_\d{2,3}$", re.IGNORECASE)
SUBJ_REGEX     = re.compile(r"^__\d{8}_\d+$")              # e.g., __20160225_17406

DRY_RUN        = False  # True = simulate, print actions, no copying
# ======================================


@dataclass
class Counters:
    copied_ok: int = 0
    failed: int = 0


@dataclass
class PlanRow:
    rel_path: str
    size: int
    mtime: float
    src: Path
    dst: Path


@dataclass
class Manifest:
    rows: Dict[str, Tuple[int, float]] = field(default_factory=dict)  # rel_path -> (size, mtime)

    @staticmethod
    def load(csv_path: Path) -> "Manifest":
        m = Manifest()
        if csv_path.exists():
            with csv_path.open('r', newline='', encoding='utf-8') as f:
                reader = csv.DictReader(f)
                # expects: rel_path,size,mtime
                for row in reader:
                    rel = row.get('rel_path', '')
                    if not rel:
                        continue
                    try:
                        size = int(float(row.get('size', '0')))
                        mtime = float(row.get('mtime', '0'))
                    except Exception:
                        continue
                    m.rows[rel] = (size, mtime)
        return m

    def save(self, csv_path: Path):
        csv_path.parent.mkdir(parents=True, exist_ok=True)
        with csv_path.open('w', newline='', encoding='utf-8') as f:
            writer = csv.DictWriter(f, fieldnames=['rel_path', 'size', 'mtime'])
            writer.writeheader()
            for rel_path, (size, mtime) in self.rows.items():
                writer.writerow({'rel_path': rel_path, 'size': size, 'mtime': mtime})

    def needs_copy(self, rel_path: str, size: int, mtime: float) -> bool:
        """Return True if file is NEW or CHANGED vs manifest."""
        if rel_path not in self.rows:
            return True
        old_size, old_mtime = self.rows[rel_path]
        return (old_size != size) or (old_mtime < mtime)

    def upsert(self, rel_path: str, size: int, mtime: float):
        self.rows[rel_path] = (size, mtime)


def _norm_rel_path(*parts: Path | str) -> str:
    """Make a stable, manifest-friendly relative path using forward slashes."""
    p = Path(*[str(x) for x in parts])
    return p.as_posix()


def list_video_files(cam_root: Path, video_exts: Iterable[str]) -> List[Path]:
    """Recursively list files under cam_root with any of the extensions (case-insensitive)."""
    out: List[Path] = []
    exts = tuple(e.lower() if e.startswith('.') else f".{e.lower()}" for e in video_exts)
    for root, _, files in os.walk(cam_root):
        rpath = Path(root)
        for fname in files:
            if fname.lower().endswith(exts):
                out.append(rpath / fname)
    return out


def discover_subjects(exp_dir: Path) -> List[Path]:
    """Find subject dirs (pattern __YYYYMMDD_XXXXX) under exp_dir/included. 
       Also supports subjects directly under exp_dir if needed."""
    candidates: List[Path] = []
    included = exp_dir / "included"
    if included.is_dir():
        for p in included.iterdir():
            if p.is_dir() and SUBJ_REGEX.match(p.name):
                candidates.append(p)
    else:
        # fallback: look one level under exp_dir (rare, but safe)
        for p in exp_dir.iterdir():
            if p.is_dir() and SUBJ_REGEX.match(p.name):
                candidates.append(p)
    return candidates


def plan_copies(multiwork_root: Path, batch_dir: Path, manifest: Manifest) -> List[PlanRow]:
    """Scan all experiments/subjects/cams and produce copy plan for new/changed files."""
    plan: List[PlanRow] = []

    experiments = [p for p in multiwork_root.iterdir() if p.is_dir() and EXP_REGEX.match(p.name)]
    print(f"[{datetime.now():%Y-%m-%d %H:%M:%S}] Scanning experiments... found {len(experiments)}")

    for exp_dir in experiments:
        subjects = discover_subjects(exp_dir)
        if not subjects:
            print(f"[WARN] No subjects found in: {exp_dir}")
            continue

        for subj_dir in subjects:
            # camera dirs directly under subject
            try:
                for cam_dir in subj_dir.iterdir():
                    if cam_dir.is_dir() and CAM_DIR_REGEX.match(cam_dir.name):
                        files = list_video_files(cam_dir, VIDEO_EXTS)
                        for src in files:
                            # rel path below cam root
                            rel_under_cam = src.relative_to(cam_dir)
                            rel_path = _norm_rel_path(exp_dir.name, subj_dir.name, cam_dir.name, rel_under_cam)

                            try:
                                stat = src.stat()
                                size = int(stat.st_size)
                                mtime = float(stat.st_mtime)
                            except FileNotFoundError:
                                continue  # file vanished mid-scan

                            if manifest.needs_copy(rel_path, size, mtime):
                                dst = batch_dir / Path(rel_path)  # preserve structure
                                plan.append(PlanRow(rel_path=rel_path, size=size, mtime=mtime, src=src, dst=dst))
            except PermissionError as e:
                print(f"[WARN] Permission error in {subj_dir}: {e}")
                continue

    # Deduplicate by rel_path (keep first occurrence)
    seen = set()
    deduped: List[PlanRow] = []
    for row in plan:
        if row.rel_path not in seen:
            seen.add(row.rel_path)
            deduped.append(row)
    return deduped


def execute_copies(plan: List[PlanRow], manifest: Manifest, dry_run: bool) -> Counters:
    cnt = Counters()
    for row in plan:
        row.dst.parent.mkdir(parents=True, exist_ok=True)
        try:
            if not dry_run:
                shutil.copy2(row.src, row.dst)  # preserves timestamps
            cnt.copied_ok += 1
            manifest.upsert(row.rel_path, row.size, row.mtime)
        except Exception as e:
            cnt.failed += 1
            print(f"[WARN] Copy failed: {row.src} -> {row.dst} ({e})")
    return cnt


def main():
    if not MULTIWORK_ROOT.is_dir():
        print(f"[ERROR] Multiwork root not found: {MULTIWORK_ROOT}")
        sys.exit(1)

    DEST_ROOT.mkdir(parents=True, exist_ok=True)
    manifest_path = DEST_ROOT / "manifest.csv"

    # single timestamped batch folder per run
    batch_dir = DEST_ROOT / datetime.now().strftime("%Y%m%d_%H%M%S")
    batch_dir.mkdir(parents=True, exist_ok=True)

    # load manifest
    manifest = Manifest.load(manifest_path)

    # plan
    plan = plan_copies(MULTIWORK_ROOT, batch_dir, manifest)
    print(f"[{datetime.now():%Y-%m-%d %H:%M:%S}] Files to copy: {len(plan)}")

    # execute
    counters = execute_copies(plan, manifest, DRY_RUN)

    # save manifest
    if not DRY_RUN:
        manifest.save(manifest_path)

    print(f"[{datetime.now():%Y-%m-%d %H:%M:%S}] Incremental backup done. "
          f"Copied OK: {counters.copied_ok}, Failed: {counters.failed}, "
          f"Batch folder: {batch_dir}")

if __name__ == "__main__":
    main()
